File: Pinball App 2/Pinball App 2/league/LeagueNextBankSupport.swift
import Foundation

func resolveLeagueNextBank(statsCSV: String?, availableBanks: Set<Int>, preferredPlayer: String?) -> Int? {
    let sortedBanks = availableBanks.sorted()
    guard !sortedBanks.isEmpty else { return nil }
    guard let statsCSV else { return sortedBanks.first }

    let statsRows = parseLeagueStatsRows(statsCSV)
    guard !statsRows.isEmpty else { return sortedBanks.first }

    let scopedRows = scopedLeagueStatsRows(statsRows, preferredPlayer: preferredPlayer)
    guard !scopedRows.isEmpty else { return sortedBanks.first }

    let latestSeason = scopedRows.map(\.season).max() ?? 0
    guard latestSeason > 0 else { return sortedBanks.first }

    let playedBanks = Set(
        scopedRows
            .filter { $0.season == latestSeason && sortedBanks.contains($0.bankNumber) }
            .map(\.bankNumber)
    )

    if let lowestMissing = sortedBanks.first(where: { !playedBanks.contains($0) }) {
        return lowestMissing
    }

    return sortedBanks.first
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/league/LeagueNextBankSupport.kt
package com.pillyliu.pinprofandroid.league

internal fun resolveNextBank(statsRows: List<StatsCsvRow>, availableBanks: Set<Int>, preferredPlayer: String?): Int? {
    val sorted = availableBanks.sorted()
    if (sorted.isEmpty()) return null
    if (statsRows.isEmpty()) return sorted.first()

    val scopedRows = scopedLeagueStatsRows(statsRows, preferredPlayer)
    if (scopedRows.isEmpty()) return sorted.first()

    val latestSeason = scopedRows.maxOfOrNull { it.season } ?: return sorted.first()
    val played = scopedRows
        .filter { it.season == latestSeason && it.bank in sorted }
        .map { it.bank }
        .toSet()

    return sorted.firstOrNull { it !in played } ?: sorted.first()
}
